File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameter;
    Solution() {
        diameter = 0;
    }

    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return diameter;
    }

    int dfs(TreeNode* node) {
        if (node == nullptr) return 0;
        int left = dfs(node->left);
        int right = dfs(node->right);
        diameter = max(diameter, left+right);
        return max(left, right)+1;
    }
};

File: Data Structures & Algorithms/climbing-stairs/submission-1.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = [0] * n
        def dfs(i: int):
            if i > n:
                return 0
            elif i == n:
                return 1
            if memo[i] != 0:
                return memo[i]

            memo[i] = dfs(i+1) + dfs(i+2)
            return memo[i]
        return dfs(0)